### ISSUE TITLE
Hover should not show docstring of inferred type but help for hover position

### ISSUE THREAD

**Issue by HedgehogCode** [python-lsp/python-lsp-server#444]
Try the following script:

```python
bar = 10
"""A variable named bar"""

bar
```

When using auto-complete while typing "bar" the documentation of the variable is resolved correctly. However, if I hover over "bar" it shows the documentation of `int()` which is not helpful at all.

The current implementation of hover uses [Script.infer](https://github.com/python-lsp/python-lsp-server/blob/develop/pylsp/plugins/hover.py#L14). Using [Script.help](https://jedi.readthedocs.io/en/latest/docs/api.html#jedi.Script.help) gives the expected hover content.

**Comment by krassowski** [comment]
Thank you, this was bugging me for a while too. If the performance-wise there is no penalty in using `help()` compared to `infer()` this is a good idea indeed.

**Comment by ccordoba12** [comment]
@HedgehogCode, could you post a screenshot of what you see when using `Script.help` on this case?

**Comment by HedgehogCode** [comment]
Sure.
Auto-complete resolve:
![Screenshot from 2023-09-20 09-06-59](https://github.com/python-lsp/python-lsp-server/assets/6631116/c4d5d125-0766-40e2-9e93-e790030d6bc9)

Hover:
![Screenshot from 2023-09-20 09-07-09](https://github.com/python-lsp/python-lsp-server/assets/6631116/26f9c6f7-a143-43f2-9a10-1e72f7532a7d)

**Comment by HedgehogCode** [comment]
`Script.infer`:
![Screenshot from 2023-09-20 09-07-09](https://github.com/python-lsp/python-lsp-server/assets/6631116/f44b8545-8c21-46ea-b15d-77189827a6fa)

`Script.help`:
![Screenshot from 2023-09-20 09-11-25](https://github.com/python-lsp/python-lsp-server/assets/6631116/b06e368c-bd34-407e-8aee-309d24a8e13b)

I did not check the complete API for Jedi yet. Therefore, I am not sure that all the post-processing of the `Script.infer` return value is also applicable for `Script.help` (in my example it just worked).

**Comment by ccordoba12** [comment]
Thanks @HedgehogCode! I agree that this would be an improvement, but it'd be better if we'd map the docstring returned by `Script.help` to its Python type so we could return instead a message saying 

> bar is an integer variable

which I think is more helpful.

Actually, I have a branch that implements a similar improvement in Spyder itself. We plan to have that in Spyder 6, but I could move that code to the server instead.

**Comment by HedgehogCode** [comment]
I agree, that this is more useful if no docstring is attached to the variable. However, if someone took the time to write a docstring for the variable I think it should be shown to the user.

Pyright for example shows both (result from a "textDocument/hover" request):
```json
{
    "contents": {
        "kind": "markdown",
        "value": "```python\n(variable) bar: Literal[10]\n```\n---\nA variable named bar"
    },
    "range": [...]
}
```

Note that the variable docstrings are not a language feature but the format is generally accepted (Jedi, sphinx, Pyright, and probably more tools support it).

**Comment by ccordoba12** [comment]
> However, if someone took the time to write a docstring for the variable I think it should be shown to the user.

Sure, that's a good point too. But I don't know how to do that and show the variable type with Jedi unless (I guess) users also add a type when declaring the variable.

**Comment by SigmaTel71** [comment]
I had started working in this direction. I found this help tooltip rather annoying than helpful, so I had to change this.
As a result, I got a functioning prototype of variable lookup tooltip. If you want to try it, it's available in my fork of pylsp. More changes coming soon!

![image.png](https://github.com/python-lsp/python-lsp-server/assets/17989678/b59c1f53-4f4f-459f-9859-142b7d700bf7)

![image.png](https://github.com/python-lsp/python-lsp-server/assets/17989678/579ec6f4-2292-4eb0-84b6-2f56f45be7d4)

**Comment by HedgehogCode** [comment]
Yes. Showing the value of a variable instead of the docstring of the built-in type is much more useful. However, I think we should also show the docstring of the variable itself if it has one (like in my initial example). I also started working in this direction [here](https://github.com/python-lsp/python-lsp-server/compare/develop...HedgehogCode:python-lsp-server:show-help-on-hover) but I want to do some more testing and provide some examples for hovering different elements.
@SigmaTel71, maybe we can combine both approaches.

**Comment by ccordoba12** [comment]
> Yes. Showing the value of a variable instead of the docstring of the built-in type is much more useful

If this is done for builtins only, then it's fine. I say it because trying to get values of objects using `repr` is a bad idea since many libraries out there have buggy or not performant `repr`'s

**Comment by SigmaTel71** [comment]
> If this is done for builtins only, then it's fine. I say it because trying to get values of objects using `repr`

I don't do that via `repr`, but via [`description`](https://github.com/python-lsp/python-lsp-server/commit/739023d4f12a3d469e89ca4ebd0a352e5f9dfb5c#diff-07f5e06788e32311b4d468256a5024fac91b3d5c0a97dd784cbaf947210ff639R56) field of `BaseName` object.

**Comment by HedgehogCode** [comment]
Note that I created the PR #452 to solve this ticket. The PR shows some screenshots of new hover results.

The proposed changes do not include the approach by @SigmaTel71 because I came to the conclusion that the definition of a statement should not be shown in the hover result. The language server protocol has a separate method to find the definition of an object: ["textDocument/definition"](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_definition). However, I am open to discuss that.

**Comment by krassowski** [comment]
> I came to the conclusion that the definition of a statement should not be shown in the hover result. The language server protocol has a separate method to find the definition of an object: "textDocument/definition". However, I am open to discuss that.

I believe this reasoning is incorrect, the intent of the definition request is not to present it in a hover but to enable go to/jump behaviour. It only returns a location object which does not have a field for text representation of the definition.

**Comment by SigmaTel71** [comment]
> However, I am open to discuss that.

I started implementing variable definition lookup on hover because this is what pyright does, but I don't want to use pyright just for that, given the fact that I moved from VSCode to Sublime Text for better workflow experience when working on Python code.
![image](https://github.com/python-lsp/python-lsp-server/assets/17989678/d309a0e2-8ee1-4cb0-90d6-57da4eeae242)
Module file from example in screenshot above is available [here](https://github.com/StranikS-Scan/WorldOfTanks-Decompiled/blob/1.22.1_EU/source/res/scripts/common/AccountCommands.py#L4).

**Comment by HedgehogCode** [comment]
> I believe this reasoning is incorrect, the intent of the definition request is not to present it in a hover but to enable go to/jump behaviour. It only returns a location object which does not have a field for text representation of the definition.

I see. That makes sense. So the question is more, if we really want to show a definition lookup in the hover result.

> I started implementing variable definition lookup on hover because this is what pyright does.

I could be wrong, but I think the pyright results are types rather than definition lookups. I put some type inference into my PR but Jedi is less strict (does not use the type `Literal[<val>]` if a variable is constant).

---

If we decide on showing a definition lookup we should answer the following questions:
- Is [`BaseName#description`](https://jedi.readthedocs.io/en/latest/docs/api-classes.html#jedi.api.classes.BaseName.description) the correct way to get the definition?
- In which cases do we want to show the definition? Only for builtin types?
- How should it be shown? Instead or next to the type information?

I am open to suggestions for enhancing my PR.

**Comment by SigmaTel71** [comment]
> * Is [`BaseName#description`](https://jedi.readthedocs.io/en/latest/docs/api-classes.html#jedi.api.classes.BaseName.description) the correct way to get the definition?

For me using `BaseName#description` looks more like a hack, because it shows the full line of code assigned to the variable, which looks more like Visual Studio's approach when you hover a macro in C/C++.
![image](https://github.com/python-lsp/python-lsp-server/assets/17989678/180ace34-09bf-4ceb-8e57-74fd6b8f8c7c)


> * In which cases do we want to show the definition? Only for builtin types?

From what I see in pyright, the definition is shown when the value is literal, no matter if it's a constant or variable. `dict` and `list` don't show its contents.
<details>
<summary>Various pyright behavior</summary>
<img src=https://github.com/python-lsp/python-lsp-server/assets/17989678/e44fc802-f39d-4083-829d-82a29bf8a168>
<img src=https://github.com/python-lsp/python-lsp-server/assets/17989678/559cee69-eadd-4dbd-a2c5-a2cc6183a37c>

Same goes with constants.

<img src=https://github.com/python-lsp/python-lsp-server/assets/17989678/71556af6-92ae-4b62-95d0-a476e17e429b>
<img src=https://github.com/python-lsp/python-lsp-server/assets/17989678/e24c2c46-0fd6-4808-8deb-72947b1e0aa8>
</details>

> * How should it be shown? Instead or next to the type information?

I think it will be effective enough to show it in order like 'type, value (if applicable), docstring (if exists)'.
